flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSink.java

```diff
@@ -172,20 +185,43 @@ public SinkRuntimeProvider getSinkRuntimeProvider(Context context) {
         final SerializationSchema<RowData> valueSerialization =
                 createSerialization(context, valueEncodingFormat, valueProjection, null);
 
-        final FlinkKafkaProducer<RowData> kafkaProducer =
-                createKafkaProducer(keySerialization, valueSerialization);
-
         if (flushMode.isEnabled() && upsertMode) {
             BufferedUpsertSinkFunction buffedSinkFunction =
```

Comment:
@AHeise Do we want to reimplement the `BufferedSinkFuntion` or extend the `KafkaSink` to offer such kind of functionality?

Reply:
I don't see the benefit of having that here `BufferedUpsertSinkFunction`. This is what Kafka producer is doing internally already. @twalthr any other opinion?

Reply:
So imho we can safely drop everything that contains producer here ;)

Reply:
>Kafka producer is doing internally already

Are you sure about that? This feature has been added recently to avoid too many tombstone messages in Kafka log. It tries to normalize the changelog by a reducing step.

Reply:
Okay the name is pretty bad. It would be awesome if we can provide the functionality as part of this PR. You'd need to implement a delegating sink where the writer is wrapped. We can also create a follow-up ticket for that.

Reply:
I opened https://issues.apache.org/jira/browse/FLINK-23735 as followup to migrate the BufferedSinkFunction to the new sink API.